mlr-org/gallery/series/2022-12-02-hyperband-subsampling/index.qmd

```diff
@@ -0,0 +1,214 @@
+If not, you should start with the [book chapter on optimization](https://mlr3book.mlr-org.com/optimization.html) or the [Hyperparameter Optimization on the Palmer Penguins Data Set](/gallery/optimization/2022-11-10-hyperparameter-optimization-on-the-palmer-penguins) post.
+A little knowledge about `r ref_pkg("mlr3pipelines")` is beneficial but not necessary to understand the example.
+
+# Hyperparameter Optimization
+
+In this post, we will optimize the hyperparameters of the support vector machine on the `r ref("mlr_tasks_sonar", "Sonar")` data set.
+We begin by constructing a classification machine by setting `type` to `"C-classification"`.
+
+```{r}
+#| message: false
+library("mlr3verse")
+
+learner = lrn("classif.svm", id = "svm", type = "C-classification")
+```
+
+The `r ref_pkg("mlr3pipelines")` package features a `r ref("PipeOp")` for subsampling.
+
+```{r}
+po("subsample")
+```
+
+The `r ref("PipeOp")` controls the size of the training data set with the `frac` parameter.
+We connect the `r ref("PipeOp")`  with the learner and get a `r ref("GraphLearner")`.
+
+```{r}
+graph_learner = as_learner(
+  po("subsample") %>>%
+  learner
+)
+```
+
+The graph learner subsamples and then fits a support vector machine on the data subset.
+The parameter set of the graph learner is a combination of the parameter sets of the `r ref("PipeOp")` and learner.
+
+```{r}
+as.data.table(graph_learner$param_set)[, .(id, lower, upper, levels)]
+```
+
+Next, we create the search space.
+We use `r ref("TuneToken")` to mark which hyperparameters should be tuned.
+We have to prefix the hyperparameters with the id of the `r ref("PipeOp", "PipeOps")`.
+The `subsample.frac` is the fidelity parameter that must be tagged with `"budget"` in the search space.
+The data set size is increased from 3.7% to 100%.
+For the other hyperparameters, we took the search space for support vector machines from the @kuehn_2018 article.
+This search space works for a wide range of data sets.
+
+```{r}
+graph_learner$param_set$set_values(
+  subsample.frac  = to_tune(p_dbl(3^-3, 1, tags = "budget")),
+  svm.kernel      = to_tune(c("linear", "polynomial", "radial")),
+  svm.cost        = to_tune(1e-4, 1e3, logscale = TRUE),
+  svm.gamma       = to_tune(1e-4, 1e3, logscale = TRUE),
+  svm.tolerance   = to_tune(1e-4, 2, logscale = TRUE),
+  svm.degree      = to_tune(2, 5)
+)
+```
+
+Support vector machines often crash or never finish the training with certain hyperparameter configurations.
+We set a timeout of 30 seconds and a fallback learner to handle these cases.
+
+```{r}
+graph_learner$encapsulate = c(train = "evaluate", predict = "evaluate")
+graph_learner$timeout = c(train = 30, predict = 30)
+graph_learner$fallback = lrn("classif.featureless")
+```
+
+Let's create the tuning instance.
+We use the `"none"` terminator because Hyperband controls the termination itself.
+
+```{r}
+instance = ti(
+  task = tsk("sonar"),
```

Comment:
why is the indentiation here different compared to the search space definition above?